cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -95,4 +110,38 @@ private String exceptionToJson(Exception ex) {
     BasicThrowable basicThrowable = new BasicThrowable(ex);
     return GSON.toJson(basicThrowable);
   }
+
+  private static class RunnableTaskBodyProducer extends BodyProducer {
+    private final InputStream responseStream;
+    private final Consumer<String> stopper;
+    private final String className;
+
+    RunnableTaskBodyProducer(InputStream responseStream, Consumer<String> stopper, String className) {
+      this.responseStream = responseStream;
+      this.stopper = stopper;
+      this.className = className;
+    }
+
+    @Override
+    public ByteBuf nextChunk() throws Exception {
+      byte[] buf = new byte[64 * 1024];
+      int len = responseStream.read(buf);
+      if (len == -1) {
+        return Unpooled.EMPTY_BUFFER;
+      }
+      return Unpooled.wrappedBuffer(buf, 0, len);
+    }
+
+    @Override
+    public void finished() throws Exception {
+      Closeables.closeQuietly(responseStream);
+      stopper.accept(className);
+    }
+
+    @Override
+    public void handleError(@Nullable Throwable cause) {
+      LOG.error("Error when sending chunks", cause);
```

Comment:
You should also close the responseStream here?

Reply:
Done!